mlflow/utils/run_data_queuing_processor.py

```diff
@@ -0,0 +1,139 @@
+
+            while len(params) > 0 or len(tags) > 0 or len(metrics) > 0:
+                try:
+                    self._processing_func(run_id=run_id, metrics=metrics, params=params, tags=tags)
+
+                    self.processed_watermark = end_watermark
+                    _logger.debug(
+                        f"run_id: {run_id}, Processed watermark: {self.processed_watermark}"
+                    )
+                    # get next batch
+                    (
+                        start_watermark,
+                        end_watermark,
+                        params,
+                        tags,
+                        metrics,
+                    ) = pending_items.get_next_batch(max_tags=100, max_metrics=250, max_params=200)
+                except Exception as e:
+                    _logger.error(
+                        f"Failed to process batches for {run_id} Start: {start_watermark} "
+                        + f" End: {end_watermark} Exception: {e}"
+                    )
+                    if self.failed_run_batches.get(run_id, None):
+                        self.failed_run_batches[run_id] = []
+                        for batch_id in range(start_watermark, end_watermark + 1):
+                            self.failed_run_batches[run_id].append(batch_id)
+
+    def _is_batch_processed(self, run_id: str, batch_id: int, timeout_sec: int = 60):
+        stop_at = time.time() + timeout_sec
+        _logger.debug(f"Start awaiting {run_id} Batch: {batch_id}")
+        while self.processed_watermark < batch_id:
+            failed_batches = self.failed_run_batches.get(run_id, {})
+            exception = failed_batches.get(batch_id, None)
+            if exception:
+                raise Exception(
+                    f"run_id: {run_id}. Failed to process batch: {batch_id} "
+                    + f"Exception: {exception!s}"
+                )
+            if time.time() > stop_at:
+                raise Exception(
+                    f"run_id: {run_id}. Batch: {batch_id} Await operation to process batch"
+                    + " timed out."
+                )
+            # Need to check better interval
+            time.sleep(5)
+        _logger.debug(f"Finished awaiting {run_id} Batch: {batch_id}")
+        return True
+
+    def log_batch_async(self, run_id, params, tags, metrics) -> RunOperations:
+        pending_run_batches = self._running_runs.get(run_id, None)
+        if not pending_run_batches:
+            pending_run_batches = self._add_run(run_id=run_id)
+
+        batch = RunBatch(params=params, tags=tags, metrics=metrics)
+        batch_id = pending_run_batches.append(batch)
+        _logger.debug(f"Enqueued watermark: {batch_id}")
+        operation_future = _RUN_BATCH_PROCESSING_STATUS_CHECK_THREADPOOL.submit(
+            self._is_batch_processed, run_id, batch_id
+        )
+        return RunOperations(operation_futures=[operation_future])
```

Comment:
do we need a wrapper class for the future? i only see one being added so directly returning it might be more ergonomic

Reply:
never mind - with the 202 metrics in AML i can see the advantage of the wrapper